### PR TITLE
Proposal class static block support

### LINKED ISSUE #43012: Proposal class static block
# Suggestion

## 🔍 Search Terms
esnext, class

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [ ] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

https://github.com/tc39/proposal-class-static-block is stage-3 。

I'm happy to implement this one if no one has already started the work(/cc: @rbuckton ).

### PR DESCRIPTION
**Pull request by Kingwl**
Closes #43308 
Fixes #43012

### CODE REVIEW

**Comment by Kingwl** [comment]
Let's try it again. @orta 

@typescript-bot pack this.

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [reply src/compiler/transformers/classFields.ts:543]: This function isn't being called because you haven't updated `doesClassElementNeedTransform`

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:979]: This function name should probably change to reflect that more than properties are supported here.

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:188]: I would add a note that this is handled by `visitClassDeclaration`.

**Review by rbuckton — CHANGES_REQUESTED** [review]
I know this is still a Draft, but thought you might appreciate some early feedback.

We should have logic in the checker to check for invalid decorators or modifiers. We also should be checking for invalid uses of `await` as an identifier in the block.

**Comment by Kingwl** [comment]
One of problem is we do not have something like `IdentifierReferenceExpression`. 

If I'm correct. `a.await` shouldn't cause an error. So we should check many thinks: 

- name of property access: `a.await`
- name of type parameter: `type await = number`
- name of type reference: `const a: await`
- etc.

Seems it's very expensive?

And we have many cases to check about the function boundary.

**Comment by rbuckton** [comment]
We could handle this via a special parser context, similar to `NodeFlags.Await` and `NodeFlags.Yield`, and then modify `isIdentifier()` in *parser.ts*.

For example, we don't allow `await` as a keyword in a type declaration in an async function today:

```ts
interface await {} // ok
async function f() {
    interface await {} // error
}
```

**Comment by rbuckton** [comment]
Also `a.await` isn't an issue since that's parsed as a property name, which allows keywords.

**Comment by rbuckton** [comment]
>  I would like to add a (flag? property?) into identifier if it's parsed by `parsePrimaryExpression`

This shouldn't be necessary.

**Comment by Kingwl** [comment]
Should we also deny the await identifiers inside type context?

**Comment by Kingwl** [comment]
Basically checked the function boundary. Computed property name is a known issue.
Please let me know if anything wrong.
Thanks!

**Comment by Kingwl** [comment]
May need to ad further tests. But I think the major semantic could be review.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
Emmmm I'm not sure why there's no errors on playground.

**Comment by orta** [comment]
Could be `target`? It seems to work like [I expect in esnext](https://www.staging-typescript.org/play?target=99&ts=4.3.0-pr-43370-14&ssl=17&ssc=3&pln=3&pc=2#code/AQYw9gdgzgLsCGwC8wCMBuAUJ0AbeUUwAwsAN47BXCzwwCWIwAZqsmlpdbQ0xdQKrhoYXAFMAdLjABzABTEJrADQklAJlWLmAZgCUXKgF9sgnoxbr26zmZh0L-QdWFRRk6fO2otG37oNBE0Mae14WHXYdLGAjIA)

**Comment by Kingwl** [comment]
Emmmm... CodeQL oom... Iet's try it.

@typescript-bot pack this.

**Comment by Kingwl** [comment]
A bit confusing:

```ts
async function foo () {
  var await = 1; // not error, but it should
  console.log(await); // error
}

async function bar () {
  let await = 2; // error
  console.log(await); // error
}
```

In chrome:

image

Why the `var` declaration doesn't parsed as an await expression? Is that working as expected?

/cc: @rbuckton

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
Thanks @orta 's help!

@typescript-bot pack this.

File: tests/baselines/reference/classStaticBlock8.errors.txt
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock8.errors.txt:12]: Just to be sure we have adequate coverage, you should also add valid `break`/`continue` cases in a `static {}`:

```ts
class C {
  static {
    outer: break outer; // valid
    loop: while (v) {
      if (v === 1) break loop; // valid
      if (v === 2) continue loop; // valid
      if (v === 3) break; // valid
      if (v === 4) continue; // valid
    }
    switch (v) {
      default: break; // valid
    }
  }
}
```

File: tests/baselines/reference/classStaticBlock18(target=es2015).js
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock18(target=es2015).js:19]: The `_b` and `_c__` variables should be scoped to the static block IIFE for the outer class expression. You may be missing a call to `startLexicalEnvironment`/`endLexicalEnvironment` in the transform somewhere.

File: tests/baselines/reference/classStaticBlock2(target=es5).js
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock2(target=es5).js:32]: Another case where there should be no collision because we're in a new function scope, this should be `a` not `a_1`.

File: tests/baselines/reference/classStaticBlock2(target=es5).js
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock2(target=es5).js:37]: Another case where there should be no collision because we're in a new function scope, this should be `a` not `a_2`.

File: tests/baselines/reference/classStaticBlock22.errors.txt
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock22.errors.txt:47]: Missing an error here. Is this what #44459 is supposed to fix?

**Review by rbuckton — CHANGES_REQUESTED** [review]
You are missing `SyntaxKind.ClassStaticBlockDeclaration` in `isDeclarationKind` in _utilitiesPublic.ts_. I'd also like to see some fourslash tests for language service functionality that might need to interact with `static {}`, such as the navigation bar, call hierarchy, and formatting.

I'd also like to see how this is handled in various module targets:
```ts
export class C {
  static x;
  static {
    C.x = 1;
  }
}
```

File: tests/baselines/reference/classStaticBlock22.errors.txt
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock22.errors.txt:47]: Yep.

Comment (rbuckton) [reply tests/baselines/reference/classStaticBlock22.errors.txt:28]: There don't seem to be any errors here after the merge. Is this something you're still working on?

Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock22.errors.txt:28]: It's unexpected. I'll to try to find out soon.

File: tests/baselines/reference/classStaticBlock4(target=es2015).errors.txt
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock4(target=es2015).errors.txt:29]: Yep. But it's working as expected for now. :XD

File: tests/baselines/reference/classStaticBlock1(target=es2015).js
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock1(target=es2015).js:18]: I'm not sure I understand correctly, just to confirm.
For blocks without private fields use IIFE. And If there are, it's still the same?

Comment (rbuckton) [reply tests/baselines/reference/classStaticBlock1(target=es2015).js:18]: With or without private fields, we only need the IIFE. That might change if `static{}` doesn't reach Stage 4 before ES2022 is ratified next year, but I think that would be unlikely.

File: tests/baselines/reference/classStaticBlock11(target=es2015).js
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock11(target=es2015).js:37]: It's static private fields's implementation. @dragomirtitian might know more about this.

Comment (rbuckton) [reply tests/baselines/reference/classStaticBlock11(target=es2015).js:37]: My point was more that we don't need to emulate the private field emit here, as it is unnecessary.

File: tests/baselines/reference/classStaticBlock22.errors.txt
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock22.errors.txt:28]: It's very strange. There's diagnostic messages at [playground](https://www.typescriptlang.org/play?alwaysStrict=false&ts=4.4.0-dev.20210622#code/IYZwngdgxgBAZgV2gFwJYHsL3emAKAShgG8AoGCmAGwFNkSZgB3YVegXxgF4TPRGIYUu1KlQkWIhQYsAI2AAnfETKUYUTCHrNW9HgEZho8dHhIoaTDHkAvZSXKUAbosYs23GIfZA). But nothing comes in test case.

File: src/compiler/checker.ts
Comment (rbuckton) [inline src/compiler/checker.ts:31309]: If you make this the first branch of the `if` statement, then you shouldn't need to exclude `containingFunctionOrClassStaticBlock.kind !== SyntaxKind.ClassStaticBlockDeclaration` on line 31301.

File: src/compiler/checker.ts
Comment (rbuckton) [inline src/compiler/checker.ts:36780]: We could probably get away with a shorter name that has a similar meaning:
```suggestion
            const container = getContainingFunctionOrClassStaticBlock(node);
```

File: src/compiler/checker.ts
Comment (rbuckton) [inline src/compiler/checker.ts:35890]:
```suggestion
            const container = getContainingFunctionOrClassStaticBlock(node);
```

File: src/compiler/checker.ts
Comment (rbuckton) [inline src/compiler/checker.ts:31275]: Can probably use a shorter name to make the `if` on line 31301 easier to read.
```suggestion
                const container = getContainingFunctionOrClassStaticBlock(node);
```

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:573]: I still don't think this is necessary. The variable goes unused.

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:559]: As far as I can tell the generated variable for the static block when private fields are present goes unused. I would just do this:

```suggestion
                const result = factory.createImmediatelyInvokedArrowFunction(statements);
```

Then we also don't need to track `classStaticBlockContainsPrivateFields`.

File: tests/baselines/reference/classStaticBlock12.js
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock12.js:22]: You can see here that `_C__` is never used, so there's no point in saving it in a local or generating the `{ value }` wrapper. The IIFE is sufficient.

File: tests/baselines/reference/classStaticBlock24(module=amd).js
Comment (rbuckton) [inline tests/baselines/reference/classStaticBlock24(module=amd).js:21]: This is what I was concerned about and why I indicated we should add the test. One of my arguments for `static {}` was that it ensured the class declaration was fully initialized before its export binding was instantiated, so that you don't see a partially initialized class in a circular import scenario. Its fine for now, since we also incorrectly emit static field initializers after the binding is instantiated. The export binding issue can be investigated in a separate PR.

**Review by rbuckton — CHANGES_REQUESTED** [review]
Thanks for all of your effort here. This is getting very close to being ready to merge.

File: tests/baselines/reference/classStaticBlock22.errors.txt
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock22.errors.txt:28]: Oh I got it. It's caused by `await;`.

You can see this [playground](https://www.typescriptlang.org/play?#code/IYZwngdgxgBAZgV2gFwJYHsL3emAKAShgG8AoGCmKTEZGYAd2FToF4YBGU8yx55ANykAvqSA).

Seems some diagnostic message will disappear when there's a `await;`;

I don't know why but it seems could be track in another thread.

Comment (rbuckton) [reply tests/baselines/reference/classStaticBlock22.errors.txt:28]: Ah, this is because we handle `await;` as a parse error in `+Await`, but `let { await }` is handled in the binder: https://github.com/microsoft/TypeScript/blob/3c71d54782789b76c1514e9dc37b42ee6a3a29fe/src/compiler/binder.ts#L2185-L2193

I would split this into two tests, one that has parse errors, and one that has static errors produced by the binder. The easiest thing to do would be to just cut out the tests that should be failing but aren't into a new test file.

**Comment by rbuckton** [comment]
If it helps, I can work on the missing language service features and tests for `static {}` so that this can be ready for 4.4 beta.

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Comment by Kingwl** [comment]
> that this can be ready for 4.4 beta.

Appreciate, if it need to be done before today end.

**Comment by rbuckton** [comment]
I'd like to get this in before noon PDT on Friday. I've pushed up support for call hierarchy, references, and go to definition along with relevant tests.

File: tests/baselines/reference/classStaticBlock4.errors.txt
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock4.errors.txt:1]: It's seems can be solved by https://github.com/microsoft/TypeScript/pull/43114.
Or we should call the iife as something like:
```ts
function C () {}
C.a = 1;
(function(){}).call(C);
```

Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock4.errors.txt:1]: Oh I got it.

File: tests/baselines/reference/classStaticBlock22.errors.txt
Comment (Kingwl) [reply tests/baselines/reference/classStaticBlock22.errors.txt:19]: Done.

**Comment by rbuckton** [comment]
From a conversation with @RyanCavanaugh, it seems like we want to try to wrap 4.4 features today. I will do what I can to get this PR merged by EOD.

**Comment by Kingwl** [comment]
I'm not sure what time zone that you say but I'll stay late this night (6/25, your morning).

**Comment by rbuckton** [comment]
I think this is good to merge except for the missing `this` support in `static {}`. Its possible we could merge this for the beta without `this` support (since referencing the class by name works), and add it once #43114 is merged. Either this PR or that one will have to deal with the merge conflicts.

**Review by rbuckton — APPROVED** [review]
I think this looks good. We can address `this` support in `static {}` in #43114.

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
After the bot checks all the things, I can merge this in the morning if there are no major breaks. Then we can work on getting this to work with #43114.

**Comment by Kingwl** [comment]
Some tests failed. But seems not my fault.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by sedghi** [comment]
Is there any way to disable this with a flag? Previously in typescript 4.5.x my esm build was 

image

But from 4.6.x I have a static block

image

Which apparently terser has problem, and also consuming apps don't recognize this

**Pull request merged**
